File: python/all_paths_from_source_to_target.py
# https://leetcode.com/problems/all-paths-from-source-to-target

# Given a directed acyclic graph (DAG) of n nodes labeled from 0 to n - 1, 
# find all possible paths from node 0 to node n - 1 and return them in any order.

# The graph is given as follows: graph[i] is a list of all nodes you 
# can visit from node i (i.e., there is a directed edge from node i to node graph[i][j]).

from typing import List

class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        end_node = len(graph) - 1
        paths = []
        def dfs(node, path):
            if node == end_node:
                paths.append(path)
            for next_node in graph[node]:
                dfs(next_node, path + [next_node])
        
        dfs(0, [0])
        return paths




if __name__ == "__main__":
    graph = [[1,2], [3], [3], []]
    print(Solution().allPathsSourceTarget(graph))
    

File: javascript/all_paths_from_source_to_target.js
// Given a directed acyclic graph (DAG) of n nodes labeled from 0 to n - 1,
// find all possible paths from node 0 to node n - 1 and return them in any order.

// The graph is given as follows: graph[i] is a list of all nodes you
// can visit from node i (i.e., there is a directed edge from node i to node graph[i][j]).

/**
 * @param {number[][]} graph
 * @return {number[][]}
 */
var allPathsSourceTarget = function (graph) {
  const results = [];
  dfs(graph, 0, [], results);
  return results;
};

function dfs(graph, node, path, results) {
  if (node === graph.length - 1) {
    results.push(path.concat(node));
    return;
  }
  for (let i = 0; i < graph[node].length; i++) {
    dfs(graph, graph[node][i], path.concat(node), results);
  }
}

console.log(allPathsSourceTarget([[1, 2], [3], [3], []]));
